Fix HLT MTV to work with singleIterPatatrack and LST procModifiers

Parsifal-2045 opened the pull request:
#### PR description:

This PR fixes `cmsDriver` recipes that try to run Validation together with HLT in a single step while using the `singleIterPatatrack` `procModifer` either alone or together with `trackingLST`. Currently, recipies such as 
```bash
cmsDriver.py step2 -s L1P2GT,HLT:75e33,VALIDATION:@hltValidation \
--processName HLTX \
--conditions auto:phase2_realistic_T33 \
--datatier GEN-SIM-DIGI-RAW,DQMIO \
--eventcontent FEVTDEBUGHLT,DQMIO \
--geometry ExtendedRun4D110 \
--era Phase2C17I13M9 \
--filein file:/eos/cms/store/relval/CMSSW_15_1_0_pre2/RelValZMM_14/GEN-SIM-DIGI-RAW/PU_150X_mcRun4_realistic_v1_STD_Run4D110_PU-v1/2580000/010145a6-c5b8-48d0-b236-0b0a1c2c3c86.root \
-n 10 \
--fileout file:step2.root \
--procModifiers alpaka,singleIterPatatrack
```
or the same with added `trackingLST` among the `procModifiers`, result in the error:
```bash
----- Begin Fatal Exception 07-May-2025 13:33:16 CEST-----------------------
An exception of category 'InvalidReference' occurred while
   [0] Processing  Event run: 1 lumi: 116 event: 11501 stream: 0
   [1] Running path 'prevalidation_step'
   [2] Calling method for module MultiTrackValidator/'hltTrackValidator'
Exception Message:
ClusterTPAssociation has OmniClusterRefs with ProductIDs 3:193,3:153 but got OmniClusterRef/ProductID with ID 2:345. This is typically caused by a configuration error.
----- End Fatal Exception -------------------------------------------------
```
which does not happen anymore after the changes.

#### PR validation:

The recipe listed above runs correctly and produces the expected plots (with or without `trackingLST`). No other difference is expected

FYI @VourMa

Parsifal-2045 committed:
Fix HLT MTV to work with singleIterPatatrack and LST procModifiers

cmsbuild commented:
A new Pull Request was created by  @Parsifal-2045 for master.

It involves the following packages:

- Validation/RecoTrack (**dqm**)


@antoniovagnerini, @cmsbuild, @rseidita can you please review it and eventually sign? Thanks.
@GiacomoSguazzoni, @VinInn, @VourMa, @dgulhan, @felicepantaleo, @missirol, @mmusich, @mtosi, @rovere, @wmtford this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mmusich commented:
mmh, aren't we testing this in the `.753` workflows?

https://github.com/cms-sw/cmssw/blob/336fa9ca3c208dc9263bcf19557c480265836c66/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py#L1841-L1852

how comes it broke without notice?

VourMa commented:
The keyword in the description is "in a single step...

> This PR fixes cmsDriver recipes that try to run Validation together with HLT in a single step

Otherwise, it runs just fine (that's how I was getting plots up till now).

mmusich commented:
> The keyword in the description is "in a single step"...

isn't this what's happening in:

```
     '-s':'DIGI:pdigi_valid,L1TrackTrigger,L1,L1P2GT,DIGI2RAW,HLT:75e33_timing,VALIDATION:@hltValidation', 
```

?

mmusich commented:
anyway @Parsifal-2045 the recipe you posted at https://github.com/cms-sw/cmssw/pull/48031#issue-3046823488 is mangled. 
Please check / correct.

VourMa commented:
> > The keyword in the description is "in a single step"...
> 
> isn't this what's happening in:
> 
> ```
>      '-s':'DIGI:pdigi_valid,L1TrackTrigger,L1,L1P2GT,DIGI2RAW,HLT:75e33_timing,VALIDATION:@hltValidation', 
> ```
> 
> ?

That's correct. One difference I notice is the change in the `--processName`. Not sure if that matters.

mmusich commented:
> That's correct. One difference I notice is the change in the --processName. Not sure if that matters.

the other thing is that the wfs .753 run the timing menu, while the recipe from Luca runs the reduced menu.

Parsifal-2045 commented:
> anyway @Parsifal-2045 the recipe you posted at [#48031 (comment)](https://github.com/cms-sw/cmssw/pull/48031#issue-3046823488) is mangled. Please check / correct.

Oops sorry, mangled copy-paste, should be fine now

mmusich commented:
test parameters:
   * workflows = 29634.753, 29634.754, 29634.755, 29634.757
   * relvals_opt = -w upgrade

mmusich commented:
@cmsbuild, please test

mmusich commented:
so apparently:
   * 29834.753 *as is* **does** run
   * 29834.753 modified to run `HLT:75e33` instead of `HLT:75e33_timing` **does** run
   * the reproducer at https://github.com/cms-sw/cmssw/pull/48031#issue-3046823488 **does not** run
   
I conclude the issue comes from not running the `DIGI:pdigi_valid,L1TrackTrigger,L1,L1P2GT,DIGI2RAW`  steps upfront but directly the `L1P2GT,HLT:75e33` step on some older `GEN-SIM-DIGI-RAW`.
I think we should add workflows to test *also* the `L1P2GT,HLT:75e33(_timing)` steps alone.

cmsbuild commented:
+1

**Size**: This PR adds an extra 24KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8ae2dc/45931/summary.html
**COMMIT**: a7fa16358e377d86b8179144e41b2c9c2c782306
**CMSSW**: CMSSW_15_1_X_2025-05-07-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48031/45931/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-07-1100+8ae2dc/68344/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 10 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-07-1100+8ae2dc/68344/validateJR.html)
 - DQMHistoTests: Total files compared: 56
 - DQMHistoTests: Total histograms compared: 4365044
 - DQMHistoTests: [Total failures: 202](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-07-1100+8ae2dc/68344/validateJR.html)
 - DQMHistoTests: Total nulls: 22
 - DQMHistoTests: Total successes: 4364800
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -5297.2970000000005 KiB( 55 files compared)
 - DQMHistoSizes: changed ( 29634.753,... ): -1764.896 KiB    HLT/Tracking
 - DQMHistoSizes: changed ( 29634.757 ): -1767.505 KiB    HLT/Tracking
 - Checked 235 log files, 202 edm output root files, 56 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-07-1100+8ae2dc/68344/triggerResults): no differences found

mmusich commented:
> DQMHistoSizes: changed ( 29634.753,... ): -1764.896 KiB HLT/Tracking
DQMHistoSizes: changed ( 29634.757 ): -1767.505 KiB HLT/Tracking

I'd like to make sure we understand these changes before signing off on this.

mmusich commented:
hold

Parsifal-2045 commented:
> > DQMHistoSizes: changed ( 29634.753,... ): -1764.896 KiB HLT/Tracking
> > DQMHistoSizes: changed ( 29634.757 ): -1767.505 KiB HLT/Tracking
> 
> I'd like to make sure we understand these changes before signing off on this.

As far as I can tell, looking the the comparison provided by the bot for the `.753 workflow` ([before](https://tinyurl.com/294l2ny3) and [after](https://tinyurl.com/28vr7y4h)) the only noticeable difference is that with these changes applied we remove the folder `hltHighPtTripletStepTrackSelectionHighPurity_hltAssociatorByHits` which was anyway filled with empty plots since `singleIterPatatrack` removes the entire `highPtTripletStep` sequence from tracking. Same thing can be said for the `.757` workflow ([before](https://tinyurl.com/23m3spxe) and [after](https://tinyurl.com/2bqhlmux))

mmusich commented:
unhold

antoniovagnerini commented:
+dqm
- spurious differences seen from unstable WF `29634.0` (open issue #45505 )
- spurious differences from unstable WF  `29834.999`(open issue #47071)

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @mandrenguyen, @rappoccio, @sextonkennedy, @antoniovilela (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
